File: uberfire-extensions/uberfire-metadata/uberfire-metadata-commons-io/src/test/java/org/uberfire/ext/metadata/io/infinispan/IndexTest.java
/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.uberfire.ext.metadata.io.infinispan;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.jboss.byteman.contrib.bmunit.BMScript;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.uberfire.ext.metadata.model.KObject;
import org.uberfire.ext.metadata.search.ClusterSegment;
import org.uberfire.ext.metadata.search.IOSearchService;
import org.uberfire.java.nio.base.FileSystemId;
import org.uberfire.java.nio.base.SegmentedPath;
import org.uberfire.java.nio.file.Path;

import static org.junit.Assert.*;

@RunWith(org.jboss.byteman.contrib.bmunit.BMUnitRunner.class)
@BMScript(dir = "byteman", value = "infinispan.btm")
public class IndexTest extends BaseIndexTest {

    @Override
    protected String[] getRepositoryNames() {
        return new String[]{getSimpleName() + "_1", getSimpleName() + "_2"};
    }

    @Test
    public void testClusterSegments() throws IOException, InterruptedException {
        setupCountDown(2);
        //Add test files
        final Path path1 = getBasePath(getSimpleName() + "_1").resolve("indexedFile1.txt");
        ioService().write(path1,
                          "content1");
        final Path path2 = getBasePath(getSimpleName() + "_2").resolve("indexedFile2.txt");
        ioService().write(path2,
                          "content2");

        //Setup ClusterSegments
        final ClusterSegment cs1 = new ClusterSegment() {
            @Override
            public String getClusterId() {
                return (((FileSystemId) getBasePath(getSimpleName() + "_1").getFileSystem()).id() + "/master");
            }

            @Override
            public String[] segmentIds() {
                return new String[]{((SegmentedPath) getBasePath(getSimpleName() + "_1")).getSegmentId()};
            }
        };
        final ClusterSegment cs2 = new ClusterSegment() {
            @Override
            public String getClusterId() {
                return ((FileSystemId) getBasePath(getSimpleName() + "_2").getFileSystem()).id() + "/master";
            }

            @Override
            public String[] segmentIds() {
                return new String[]{((SegmentedPath) getBasePath(getSimpleName() + "_2")).getSegmentId()};
            }
        };

        waitForCountDown(10000);

        final Map<String, Object> attributes = new HashMap<String, Object>() {{
            put("filename",
                "*txt");
        }};

        //Attribute Search
        {
            final int hits = config.getSearchIndex().searchByAttrsHits(attributes);
            final List<KObject> results = config.getSearchIndex().searchByAttrs(attributes,
                                                                                new IOSearchService.NoOpFilter());
            assertEquals(0,
                         hits);
            assertEquals(0,
                         results.size());
        }

        {
            final int hits = config.getSearchIndex().searchByAttrsHits(attributes,
                                                                       cs1);
            final List<KObject> results = config.getSearchIndex().searchByAttrs(attributes,
                                                                                new IOSearchService.NoOpFilter(),
                                                                                cs1);
            assertEquals(1,
                         hits);
            assertEquals(1,
                         results.size());
        }

        {
            final int hits = config.getSearchIndex().searchByAttrsHits(attributes,
                                                                       cs2);
            final List<KObject> results = config.getSearchIndex().searchByAttrs(attributes,
                                                                                new IOSearchService.NoOpFilter(),
                                                                                cs2);
            assertEquals(1,
                         hits);
            assertEquals(1,
                         results.size());
        }

        {
            final int hits = config.getSearchIndex().searchByAttrsHits(attributes,
                                                                       cs1,
                                                                       cs2);
            final List<KObject> results = config.getSearchIndex().searchByAttrs(attributes,
                                                                                new IOSearchService.NoOpFilter(),
                                                                                cs1,
                                                                                cs2);
            assertEquals(2,
                         hits);
            assertEquals(2,
                         results.size());
        }

        //Full Text Search
        {
            final int hits = config.getSearchIndex().fullTextSearchHits("*indexed*");
            final List<KObject> results = config.getSearchIndex().fullTextSearch("*indexed*",
                                                                                 new IOSearchService.NoOpFilter());
            assertEquals(0,
                         hits);
            assertEquals(0,
                         results.size());
        }

        {
            final int hits = config.getSearchIndex().fullTextSearchHits("*indexed*",
                                                                        cs1);
            final List<KObject> results = config.getSearchIndex().fullTextSearch("*indexed*",
                                                                                 new IOSearchService.NoOpFilter(),
                                                                                 cs1);
            assertEquals(1,
                         hits);
            assertEquals(1,
                         results.size());
        }

        {
            final int hits = config.getSearchIndex().fullTextSearchHits("*indexed*",
                                                                        cs2);
            final List<KObject> results = config.getSearchIndex().fullTextSearch("*indexed*",
                                                                                 new IOSearchService.NoOpFilter(),
                                                                                 cs2);
            assertEquals(1,
                         hits);
            assertEquals(1,
                         results.size());
        }

        {
            final int hits = config.getSearchIndex().fullTextSearchHits("*indexed*",
                                                                        cs1,
                                                                        cs2);
            final List<KObject> results = config.getSearchIndex().fullTextSearch("*indexed*",
                                                                                 new IOSearchService.NoOpFilter(),
                                                                                 cs1,
                                                                                 cs2);
            assertEquals(2,
                         hits);
            assertEquals(2,
                         results.size());
        }
    }

    protected String getSimpleName() {
        return "indexable_" + IndexTest.this.getClass().getSimpleName();
    }
}

File: packages/serverless-logic-sandbox/src/__tests__/extension/index.test.ts
/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Minimatch } from "minimatch";
import { GLOB_PATTERN } from "../../extension";

describe("extension", () => {
  describe("GLOB_PATTERN", () => {
    describe("all", () => {
      it.each([[true, "anything", "foo.txt"]])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.all, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });

    describe("allExceptDockerfiles", () => {
      it.each([
        [false, "a Dockerfile in the root", "Dockerfile"],
        [false, "a Dockerfile in a folder", "bar/Dockerfile"],
        [false, "a .dockerignore in the root", ".dockerignore"],
        [false, "a .dockerignore in a folder", "bar/.dockerignore"],
        [true, "not a docker file in the root", "foo.txt"],
        [true, "not a docker file in a folder", "bar/foo.txt"],
      ])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.allExceptDockerfiles, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });

    describe("sw", () => {
      it.each([
        [false, "not a serverless workflow file in the root", "foo.txt"],
        [false, "not a serverless workflow file in the root", "foo.sw"],
        [false, "json file in the root", "foo.json"],
        [false, "yaml file in the root", "foo.yaml"],
        [false, "yml file in the root", "foo.yml"],
        [false, "json file in the root", "foo.json"],
        [true, "sw json file in the root", "foo.sw.json"],
        [true, "sw yaml file in the root", "foo.sw.yaml"],
        [true, "sw yml file in the root", "foo.sw.yml"],
        [true, "sw json file in a folder", "bar/foo.sw.json"],
        [true, "sw yaml file in a folder", "bar/foo.sw.yaml"],
        [true, "sw yml file in a folder", "bar/foo.sw.yml"],
      ])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.sw, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });

    describe("yard", () => {
      it.each([
        [false, "not a serverless decision file in the root", "foo.txt"],
        [false, "not a serverless decision file in the root", "foo.yard"],
        [false, "json file in the root", "foo.json"],
        [false, "yaml file in the root", "foo.yaml"],
        [false, "yml file in the root", "foo.yml"],
        [false, "json file in the root", "foo.json"],
        [true, "yard json file in the root", "foo.yard.json"],
        [true, "yard yaml file in the root", "foo.yard.yaml"],
        [true, "yard yml file in the root", "foo.yard.yml"],
        [true, "yard json file in a folder", "bar/foo.yard.json"],
        [true, "yard yaml file in a folder", "bar/foo.yard.yaml"],
        [true, "yard yml file in a folder", "bar/foo.yard.yml"],
      ])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.yard, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });

    describe("dash", () => {
      it.each([
        [false, "not a dashboard file in the root", "foo.txt"],
        [false, "not a dashboard file in the root", "foo.dash"],
        [false, "json file in the root", "foo.json"],
        [false, "yaml file in the root", "foo.yaml"],
        [false, "yml file in the root", "foo.yml"],
        [false, "json file in the root", "foo.json"],
        [false, "dash json file in the root", "foo.dash.json"],
        [false, "dash json file in a folder", "bar/foo.dash.json"],
        [true, "dash yaml file in the root", "foo.dash.yaml"],
        [true, "dash yml file in the root", "foo.dash.yml"],
        [true, "dash yaml file in a folder", "bar/foo.dash.yaml"],
        [true, "dash yml file in a folder", "bar/foo.dash.yml"],
      ])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.dash, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });

    describe("spec", () => {
      it.each([
        [false, "not a spec file in the root", "foo.txt"],
        [false, "not a spec file in the root", "foo.spec"],
        [false, "json file in the root", "foo.json"],
        [false, "yaml file in the root", "foo.yaml"],
        [false, "yml file in the root", "foo.yml"],
        [false, "json file in the root", "foo.json"],
        [true, "spec json file in the root", "foo.spec.json"],
        [true, "spec yaml file in the root", "foo.spec.yaml"],
        [true, "spec yml file in the root", "foo.spec.yml"],
        [true, "spec json file in a folder", "bar/foo.spec.json"],
        [true, "spec yaml file in a folder", "bar/foo.spec.yaml"],
        [true, "spec yml file in a folder", "bar/foo.spec.yml"],
        [true, "specs json file in a folder", "bar/foo.specs.json"],
        [true, "specs yaml file in a folder", "bar/foo.specs.yaml"],
        [true, "specs yml file in a folder", "bar/foo.specs.yml"],
        [true, "spec json file in the root", "spec.json"],
        [true, "spec yaml file in the root", "spec.yaml"],
        [true, "spec yml file in the root", "spec.yml"],
        [true, "spec json file in a folder", "bar/spec.json"],
        [true, "spec yaml file in a folder", "bar/spec.yaml"],
        [true, "spec yml file in a folder", "bar/spec.yml"],
        [true, "specs json file in a folder", "bar/specs.json"],
        [true, "specs yaml file in a folder", "bar/specs.yaml"],
        [true, "specs yml file in a folder", "bar/specs.yml"],
      ])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.spec, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });

    describe("sw_spec", () => {
      it.each([
        [false, "not a spec or serverless workflow file in the root", "foo.txt"],
        [false, "not a spec  or serverless workflow file in the root", "foo.spec"],
        [false, "not a spec  or serverless workflow file in the root", "foo.sw"],
        [false, "json file in the root", "foo.json"],
        [false, "yaml file in the root", "foo.yaml"],
        [false, "yml file in the root", "foo.yml"],
        [false, "json file in the root", "foo.json"],
        [true, "spec json file in the root", "foo.spec.json"],
        [true, "spec yaml file in the root", "foo.spec.yaml"],
        [true, "spec yml file in the root", "foo.spec.yml"],
        [true, "spec json file in a folder", "bar/foo.spec.json"],
        [true, "spec yaml file in a folder", "bar/foo.spec.yaml"],
        [true, "spec yml file in a folder", "bar/foo.spec.yml"],
        [true, "specs json file in a folder", "bar/foo.specs.json"],
        [true, "specs yaml file in a folder", "bar/foo.specs.yaml"],
        [true, "specs yml file in a folder", "bar/foo.specs.yml"],
        [true, "spec json file in the root", "spec.json"],
        [true, "spec yaml file in the root", "spec.yaml"],
        [true, "spec yml file in the root", "spec.yml"],
        [true, "spec json file in a folder", "bar/spec.json"],
        [true, "spec yaml file in a folder", "bar/spec.yaml"],
        [true, "spec yml file in a folder", "bar/spec.yml"],
        [true, "specs json file in a folder", "bar/specs.json"],
        [true, "specs yaml file in a folder", "bar/specs.yaml"],
        [true, "specs yml file in a folder", "bar/specs.yml"],
        [true, "sw json file in the root", "foo.sw.json"],
        [true, "sw yaml file in the root", "foo.sw.yaml"],
        [true, "sw yml file in the root", "foo.sw.yml"],
        [true, "sw json file in a folder", "bar/foo.sw.json"],
        [true, "sw yaml file in a folder", "bar/foo.sw.yaml"],
        [true, "sw yml file in a folder", "bar/foo.sw.yml"],
      ])("should be %s when path is %s", (result, _desc, path) => {
        const matcher = new Minimatch(GLOB_PATTERN.sw_spec, { dot: true });
        expect(matcher.match(path)).toBe(result);
      });
    });
  });
});
